## 0x84 线程通信

### 方法

- `void wait()` 让当前线程等待并释放所占锁，知道另一个线程调用 `notify()` 方法或 `notifyAll()` 方法
- `void notify()` 唤醒正在等待的单个线程
- `void notifyAll()` 唤醒正在等待的所有线程

#### 注意

- 上述方法应该使用当前同步锁对象进行调用

### 案例：生产者消费者

- 3个生产者线程，负责生产包子，每个线程每次只能生产1个包子放在桌子上
- 2个消费者线程负责吃包子，每人每次只能从桌子上拿1个包子吃

```java
// Main.java
public class Main {
    public static void main(String[] args) {
        Desk desk = new Desk();

        // 3个厨师
        new Thread(() -> {
            while (true) {
                desk.put();
            }
        }, "厨师1").start();

        new Thread(() -> {
            while (true) {
                desk.put();
            }
        }, "厨师2").start();

        new Thread(() -> {
            while (true) {
                desk.put();
            }
        }, "厨师3").start();

        // 2个吃货

        new Thread(() -> {
            while (true) {
                desk.get();
            }
        }, "吃货1").start();

        new Thread(() -> {
            while (true) {
                desk.get();
            }
        }, "吃货2").start();
    }
}

// Desk.java
public class Desk {
    private List<String> list = new ArrayList<>();

    // 放1个包子
    public synchronized void put() {
        String name = Thread.currentThread().getName();
        // 判断是否有包子
        try {
            if (list.isEmpty()) {
                list.add(name + "做的肉包子");
                System.out.println(name + "做了一个肉包子");
                Thread.sleep(2000);
            }

            // 等待别人，唤醒自己
            this.notifyAll();
            this.wait();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    // 取1个包子
    public synchronized void get() {
        String name = Thread.currentThread().getName();
        try {
            if (list.size() == 1) {
                // 有包子，吃了
                System.out.println(name + "吃了" + list.get(0));
                list.clear();
                Thread.sleep(1000);
            }

            // 等待别人，唤醒自己
            this.notifyAll();
            this.wait();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
```